Handle http status codes

garethbowen opened the pull request:
Closes #

### I have verified this PR works in these browsers (latest versions):

- [ ] Chrome
- [x] Firefox
- [ ] Safari (macOS)
- [ ] Safari (iOS)
- [ ] Chrome for Android
- [ ] Not applicable

### What else has been done to verify that this works as intended?

### Why is this the best possible solution? Were any other approaches considered?

I'm not sure about outputting the error code but because it's a demo app I'm assuming it's for more technical users who would appreciate this.

### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

None. Provides more info about errors.

### Do we need any specific form for testing your changes? If so, please attach one.

None.

### What's changed

Before this change if the request to the XLSForm Online service failed the status code was ignored and the server attempted to parse the JSON response which doesn't work. This shows a misleading error message which implied (to me) that there was something wrong with the input rather than checking that the service was running.

Before:

<img width="513" height="375" alt="Screenshot From 2025-08-19 10-51-57" src="https://github.com/user-attachments/assets/860b25cb-500e-47aa-a2ce-db29a40324ee" />

After:

<img width="513" height="375" alt="Screenshot From 2025-08-19 10-55-26" src="https://github.com/user-attachments/assets/d28bc0b0-99d4-4fcd-ab7a-a18c5afa4dd6" />

I also changed the code from a mix of async/await and promises, to be completely async/await which I think is easier to read, but I did this in a separate commit so I can revert that easily if you just want the bug fix and nothing else.

latin-panda reviewed (commented):
This looks great, thanks! The demo is automatically deployed at https://getodk.org/web-forms-preview/ once the changes are merged.

There was a bug in the CI that prevented it from running for pull requests from forks. Could you rebase to include the latest fix? Then we will see the CI checks passing, and I'll approve the PR.

By the way, how did you encounter that error? I attempted to stop the XLSForm Online service, disable the network, and restart it with incorrect configurations, but it always ends up in the catch block.

garethbowen commented:
I've rebased - awaiting approval.

I got the error while I was figuring out how to get CORS working. In one misguided attempt I changed the port that the demo was running on to 8000 which was the port that the xlsform service was configured to run on. Because the ports were the same the CORS issue went away, but the request went to the demo which didn't have the requested endpoint and returned a 404. While this is probably an edge case not worth worrying about, I expect it would be possible to get the same result with any response from the xlsform service with invalid JSON (eg: empty response).

You will still end up in the catch block because the `res.json()` call throws an Error, the problem is the error message isn't very helpful (see the before screenshot).

latin-panda reviewed (approved):
Thank you! ✨

There are a couple of lint errors. The format error is usually fixed by running `yarn format` in the root folder. Since this project is quite strict with types, we might want to define something to describe the returned object. That would make it ready to squash and merge.

garethbowen commented:
Same here - should pass now!
